core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1610,6 +1566,16 @@ class Log(@volatile var dir: File,
    */
   def addSegment(segment: LogSegment) = this.segments.put(segment.baseOffset, segment)
 
+  private def maybeHandleIOException[T](msg: => String)(fun: => T): T = {
```

Comment:
I have thought about this. We can do it if we add another 1-2 parameters to this method. This is because both the object `logDirFailureChannel` and `dir` in the body of `maybeHandleIOException` belong to `Log`. Thus the `maybeHandleIOException()` in its current form can not be static.

It seems to me that the code is simpler by putting it only in Log. We can move it to CoreUtils with the extra parameter if you think that is better.